Extension interfaces with generics are causing crashes

I am trying to write a blog post about the usage of the Whiteboard Implementation for Jakarta RESTful Web Services. I want it to be similar to my old [Build REST services with OSGi JAX-RS whiteboard](https://vogella.com/blog/build-rest-services-with-osgi-jax-rs-whiteboard/).

I tried the example by using the Jetty 11 bundles directly instead of the Apache Felix Http Jetty bundle, and with the Apache Felix Http Jetty bundle. In both cases this works fine for a simple service. But when I try to add a converter for JSON, I get the following exception:

```
Jan. 30, 2024 4:15:27 PM org.glassfish.jersey.server.ServerRuntime$Responder process
WARNUNG: An exception mapping did not successfully produce and processed a response. Logging the exception propagated to the default exception mapper.
java.lang.IllegalStateException: ServiceLocatorImpl(__HK2_Generated_1,1,822606505) has been shut down
	at org.jvnet.hk2.internal.ServiceLocatorImpl.checkState(ServiceLocatorImpl.java:2399)
	at org.jvnet.hk2.internal.ServiceLocatorImpl.internalGetService(ServiceLocatorImpl.java:738)
	at org.jvnet.hk2.internal.ServiceLocatorImpl.internalGetService(ServiceLocatorImpl.java:732)
	at org.jvnet.hk2.internal.ServiceLocatorImpl.getService(ServiceLocatorImpl.java:702)
	at org.glassfish.jersey.inject.hk2.AbstractHk2InjectionManager.getInstance(AbstractHk2InjectionManager.java:160)
	at org.glassfish.jersey.inject.hk2.ImmediateHk2InjectionManager.getInstance(ImmediateHk2InjectionManager.java:30)
	at org.glassfish.jersey.server.ApplicationHandler.lambda$initialize$4(ApplicationHandler.java:424)
	at org.glassfish.jersey.server.internal.process.ReferencesInitializer.apply(ReferencesInitializer.java:57)
	at org.glassfish.jersey.server.internal.process.ReferencesInitializer.apply(ReferencesInitializer.java:31)
	at org.glassfish.jersey.process.internal.Stages$LinkedStage.apply(Stages.java:284)
	at org.glassfish.jersey.process.internal.Stages.process(Stages.java:173)
	at org.glassfish.jersey.server.ServerRuntime$1.run(ServerRuntime.java:253)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:248)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:244)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:292)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:274)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:244)
	at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:265)
	at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:240)
	at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:697)
	at org.glassfish.jersey.servlet.WebComponent.serviceImpl(WebComponent.java:394)
	at org.glassfish.jersey.servlet.WebComponent.service(WebComponent.java:346)
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:357)
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:311)
	at org.eclipse.osgitech.rest.runtime.WhiteboardServletContainer.service(WhiteboardServletContainer.java:137)
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:205)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:764)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:529)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:221)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1380)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:176)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:484)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:174)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1302)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:129)
	at org.eclipse.jetty.server.handler.HandlerList.handle(HandlerList.java:51)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:122)
	at org.eclipse.jetty.server.Server.handle(Server.java:563)
	at org.eclipse.jetty.server.HttpChannel.lambda$handle$0(HttpChannel.java:505)
	at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:762)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:497)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:282)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:314)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:100)
	at org.eclipse.jetty.io.SelectableChannelEndPoint$1.run(SelectableChannelEndPoint.java:53)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:934)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1078)
	at java.base/java.lang.Thread.run(Thread.java:840)

Jan. 30, 2024 4:15:27 PM org.glassfish.jersey.server.ServerRuntime$Responder writeResponse
SCHWERWIEGEND: Error while closing the output stream in order to commit response.
org.glassfish.jersey.server.ContainerException: java.util.concurrent.ExecutionException: java.lang.IllegalStateException: ServiceLocatorImpl(__HK2_Generated_1,1,822606505) has been shut down
	at org.glassfish.jersey.servlet.internal.ResponseWriter.getResponseContext(ResponseWriter.java:278)
	at org.glassfish.jersey.servlet.internal.ResponseWriter.writeResponseStatusAndHeaders(ResponseWriter.java:128)
	at org.glassfish.jersey.server.ServerRuntime$Responder$1.getOutputStream(ServerRuntime.java:667)
	at org.glassfish.jersey.message.internal.CommittingOutputStream.commitStream(CommittingOutputStream.java:171)
	at org.glassfish.jersey.message.internal.CommittingOutputStream.flushBuffer(CommittingOutputStream.java:276)
	at org.glassfish.jersey.message.internal.CommittingOutputStream.commit(CommittingOutputStream.java:232)
	at org.glassfish.jersey.message.internal.CommittingOutputStream.close(CommittingOutputStream.java:247)
	at org.glassfish.jersey.message.internal.OutboundMessageContext.close(OutboundMessageContext.java:865)
	at org.glassfish.jersey.server.ContainerResponse.close(ContainerResponse.java:403)
	at org.glassfish.jersey.server.ServerRuntime$Responder.writeResponse(ServerRuntime.java:749)
	at org.glassfish.jersey.server.ServerRuntime$Responder.processResponse(ServerRuntime.java:385)
	at org.glassfish.jersey.server.ServerRuntime$Responder.processResponseWithDefaultExceptionMapper(ServerRuntime.java:633)
	at org.glassfish.jersey.server.ServerRuntime$Responder.process(ServerRuntime.java:458)
	at org.glassfish.jersey.server.ServerRuntime$1.run(ServerRuntime.java:269)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:248)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:244)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:292)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:274)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:244)
	at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:265)
	at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:240)
	at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:697)
	at org.glassfish.jersey.servlet.WebComponent.serviceImpl(WebComponent.java:394)
	at org.glassfish.jersey.servlet.WebComponent.service(WebComponent.java:346)
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:357)
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:311)
	at org.eclipse.osgitech.rest.runtime.WhiteboardServletContainer.service(WhiteboardServletContainer.java:137)
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:205)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:764)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:529)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:221)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1380)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:176)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:484)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:174)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1302)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:129)
	at org.eclipse.jetty.server.handler.HandlerList.handle(HandlerList.java:51)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:122)
	at org.eclipse.jetty.server.Server.handle(Server.java:563)
	at org.eclipse.jetty.server.HttpChannel.lambda$handle$0(HttpChannel.java:505)
	at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:762)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:497)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:282)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:314)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:100)
	at org.eclipse.jetty.io.SelectableChannelEndPoint$1.run(SelectableChannelEndPoint.java:53)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:934)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1078)
	at java.base/java.lang.Thread.run(Thread.java:840)
Caused by: java.util.concurrent.ExecutionException: java.lang.IllegalStateException: ServiceLocatorImpl(__HK2_Generated_1,1,822606505) has been shut down
	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:396)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2073)
	at org.glassfish.jersey.servlet.internal.ResponseWriter.getResponseContext(ResponseWriter.java:276)
	... 49 more
Caused by: java.lang.IllegalStateException: ServiceLocatorImpl(__HK2_Generated_1,1,822606505) has been shut down
	at org.jvnet.hk2.internal.ServiceLocatorImpl.checkState(ServiceLocatorImpl.java:2399)
	at org.jvnet.hk2.internal.ServiceLocatorImpl.internalGetService(ServiceLocatorImpl.java:738)
	at org.jvnet.hk2.internal.ServiceLocatorImpl.internalGetService(ServiceLocatorImpl.java:732)
	at org.jvnet.hk2.internal.ServiceLocatorImpl.getService(ServiceLocatorImpl.java:702)
	at org.glassfish.jersey.inject.hk2.AbstractHk2InjectionManager.getInstance(AbstractHk2InjectionManager.java:160)
	at org.glassfish.jersey.inject.hk2.ImmediateHk2InjectionManager.getInstance(ImmediateHk2InjectionManager.java:30)
	at org.glassfish.jersey.server.ApplicationHandler.lambda$initialize$4(ApplicationHandler.java:424)
	at org.glassfish.jersey.server.internal.process.ReferencesInitializer.apply(ReferencesInitializer.java:57)
	at org.glassfish.jersey.server.internal.process.ReferencesInitializer.apply(ReferencesInitializer.java:31)
	at org.glassfish.jersey.process.internal.Stages$LinkedStage.apply(Stages.java:284)
	at org.glassfish.jersey.process.internal.Stages.process(Stages.java:173)
	at org.glassfish.jersey.server.ServerRuntime$1.run(ServerRuntime.java:253)
	... 36 more

[qtp631911774-76] WARN org.eclipse.jetty.server.HttpChannel - /modify/schnulli
jakarta.servlet.ServletException: java.lang.IllegalStateException: ServiceLocatorImpl(__HK2_Generated_1,1,822606505) has been shut down
	at org.glassfish.jersey.servlet.WebComponent.serviceImpl(WebComponent.java:409)
	at org.glassfish.jersey.servlet.WebComponent.service(WebComponent.java:346)
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:357)
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:311)
	at org.eclipse.osgitech.rest.runtime.WhiteboardServletContainer.service(WhiteboardServletContainer.java:137)
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:205)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:764)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:529)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:221)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1380)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:176)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:484)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:174)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1302)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:129)
	at org.eclipse.jetty.server.handler.HandlerList.handle(HandlerList.java:51)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:122)
	at org.eclipse.jetty.server.Server.handle(Server.java:563)
	at org.eclipse.jetty.server.HttpChannel.lambda$handle$0(HttpChannel.java:505)
	at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:762)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:497)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:282)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:314)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:100)
	at org.eclipse.jetty.io.SelectableChannelEndPoint$1.run(SelectableChannelEndPoint.java:53)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:934)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1078)
	at java.base/java.lang.Thread.run(Thread.java:840)
Caused by: java.lang.IllegalStateException: ServiceLocatorImpl(__HK2_Generated_1,1,822606505) has been shut down
	at org.jvnet.hk2.internal.ServiceLocatorImpl.checkState(ServiceLocatorImpl.java:2399)
	at org.jvnet.hk2.internal.ServiceLocatorImpl.internalGetService(ServiceLocatorImpl.java:738)
	at org.jvnet.hk2.internal.ServiceLocatorImpl.internalGetService(ServiceLocatorImpl.java:732)
	at org.jvnet.hk2.internal.ServiceLocatorImpl.getService(ServiceLocatorImpl.java:702)
	at org.glassfish.jersey.inject.hk2.AbstractHk2InjectionManager.getInstance(AbstractHk2InjectionManager.java:160)
	at org.glassfish.jersey.inject.hk2.ImmediateHk2InjectionManager.getInstance(ImmediateHk2InjectionManager.java:30)
	at org.glassfish.jersey.server.ApplicationHandler.lambda$initialize$4(ApplicationHandler.java:424)
	at org.glassfish.jersey.server.internal.process.ReferencesInitializer.apply(ReferencesInitializer.java:57)
	at org.glassfish.jersey.server.internal.process.ReferencesInitializer.apply(ReferencesInitializer.java:31)
	at org.glassfish.jersey.process.internal.Stages$LinkedStage.apply(Stages.java:284)
	at org.glassfish.jersey.process.internal.Stages.process(Stages.java:173)
	at org.glassfish.jersey.server.ServerRuntime$1.run(ServerRuntime.java:253)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:248)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:244)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:292)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:274)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:244)
	at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:265)
	at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:240)
	at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:697)
	at org.glassfish.jersey.servlet.WebComponent.serviceImpl(WebComponent.java:394)
	... 27 more
[qtp631911774-76] WARN org.eclipse.jetty.server.HttpChannelState - unhandled due to prior sendError
jakarta.servlet.ServletException: java.lang.IllegalStateException: ServiceLocatorImpl(__HK2_Generated_1,1,822606505) has been shut down
	at org.glassfish.jersey.servlet.WebComponent.serviceImpl(WebComponent.java:409)
	at org.glassfish.jersey.servlet.WebComponent.service(WebComponent.java:346)
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:357)
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:311)
	at org.eclipse.osgitech.rest.runtime.WhiteboardServletContainer.service(WhiteboardServletContainer.java:137)
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:205)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:764)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:529)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:221)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1380)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:176)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:484)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:174)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1302)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:129)
	at org.eclipse.jetty.server.handler.HandlerList.handle(HandlerList.java:51)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:122)
	at org.eclipse.jetty.server.Server.handle(Server.java:563)
	at org.eclipse.jetty.server.HttpChannel.lambda$handle$0(HttpChannel.java:505)
	at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:762)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:497)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:282)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:314)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:100)
	at org.eclipse.jetty.io.SelectableChannelEndPoint$1.run(SelectableChannelEndPoint.java:53)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:934)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1078)
	at java.base/java.lang.Thread.run(Thread.java:840)
Caused by: java.lang.IllegalStateException: ServiceLocatorImpl(__HK2_Generated_1,1,822606505) has been shut down
	at org.jvnet.hk2.internal.ServiceLocatorImpl.checkState(ServiceLocatorImpl.java:2399)
	at org.jvnet.hk2.internal.ServiceLocatorImpl.internalGetService(ServiceLocatorImpl.java:738)
	at org.jvnet.hk2.internal.ServiceLocatorImpl.internalGetService(ServiceLocatorImpl.java:732)
	at org.jvnet.hk2.internal.ServiceLocatorImpl.getService(ServiceLocatorImpl.java:702)
	at org.glassfish.jersey.inject.hk2.AbstractHk2InjectionManager.getInstance(AbstractHk2InjectionManager.java:160)
	at org.glassfish.jersey.inject.hk2.ImmediateHk2InjectionManager.getInstance(ImmediateHk2InjectionManager.java:30)
	at org.glassfish.jersey.server.ApplicationHandler.lambda$initialize$4(ApplicationHandler.java:424)
	at org.glassfish.jersey.server.internal.process.ReferencesInitializer.apply(ReferencesInitializer.java:57)
	at org.glassfish.jersey.server.internal.process.ReferencesInitializer.apply(ReferencesInitializer.java:31)
	at org.glassfish.jersey.process.internal.Stages$LinkedStage.apply(Stages.java:284)
	at org.glassfish.jersey.process.internal.Stages.process(Stages.java:173)
	at org.glassfish.jersey.server.ServerRuntime$1.run(ServerRuntime.java:253)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:248)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:244)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:292)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:274)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:244)
	at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:265)
	at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:240)
	at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:697)
	at org.glassfish.jersey.servlet.WebComponent.serviceImpl(WebComponent.java:394)
	... 27 more
```

I tried various configurations, added bundles explicitly, but nothing works.

On the one side I noticed that Eclipse Parsson is still used in version 1.1.2, but from the changelog it seems that several OSGi related issues were resolved in the meantime. Not sure if this is related, but I opened PR #38 to update the dependency.

Another thing I noticed is that there are many configuration reload executions. At least two of them look strange codewise.

In `JerseyServiceRuntime` line 514 a reload is triggered if `JerseyApplicationProvider#isChanged(Application)` returns true. The interesting thing is, that method always returns true. And that method is not overriding a parent class method, and even the javadoc looks weird. I wonder if this is intended. But of course I am also not sure if this is the cause for the issue.

My example is available in GitHub: https://github.com/fipro78/osgi-jakartars

The **main** branch contains the working example with the simple service setup.
The **json** branch contains the extension with the JSON handling, which shows the error on requesting the resource in a browser.

Once the app is started via the _app.bndrun_  file, you can try to access the resource via [http://localhost:8080/modify/fubar](http://localhost:8080/modify/fubar), which shows the error in the console.

--- a/org.eclipse.osgitech.rest/src/main/java/org/eclipse/osgitech/rest/proxy/ExtensionProxyFactory.java
+++ b/org.eclipse.osgitech.rest/src/main/java/org/eclipse/osgitech/rest/proxy/ExtensionProxyFactory.java
@@ -17,12 +17,19 @@ import java.lang.annotation.Annotation;
 import java.lang.reflect.InvocationTargetException;
 import java.lang.reflect.Method;
 import java.lang.reflect.ParameterizedType;
+import java.lang.reflect.TypeVariable;
+import java.lang.reflect.WildcardType;
+import java.util.ArrayList;
 import java.util.Arrays;
+import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
+import java.util.Set;
 import java.util.function.Function;
+import java.util.function.Predicate;
 import java.util.function.Supplier;
 import java.util.stream.Collectors;
+import java.util.stream.Stream;
 
 import org.objectweb.asm.AnnotationVisitor;
 import org.objectweb.asm.ClassWriter;
@@ -54,13 +61,11 @@ public class ExtensionProxyFactory {
 	 * @param simpleName
 	 */
 	public static byte[] generateClass(String className, Object delegate, List<Class<?>> contracts) {
-		Map<String, ParameterizedType> typeInfo = Arrays.stream(delegate.getClass().getGenericInterfaces())
-				.filter(ParameterizedType.class::isInstance)
-				.map(ParameterizedType.class::cast)
-				.collect(Collectors.toMap(i -> i.getRawType().getTypeName(), Function.identity()));
+		Class<? extends Object> delegateClazz = delegate.getClass();
+		Map<String, ParameterizedType> typeInfo = getInterfacesAndGenericSuperclasses(delegateClazz, new ArrayList<>(), new HashSet<>(contracts))
+				.stream().collect(Collectors.toMap(i -> i.getRawType().getTypeName(), Function.identity()));
 		
-		
-		String sig = generateGenericClassSignature(typeInfo, contracts);
+		String sig = generateGenericClassSignature(delegateClazz, typeInfo, contracts);
 		
 		
 		try {
@@ -73,7 +78,7 @@ public class ExtensionProxyFactory {
 						.map(Type::getInternalName)
 						.toArray(String[]::new));
 			
-			for (Annotation annotation : delegate.getClass().getAnnotations()) {
+			for (Annotation annotation : delegateClazz.getAnnotations()) {
 				AnnotationVisitor av = cw.visitAnnotation(Type.getDescriptor(annotation.annotationType()), true);
 				visitAnnotationMembers(annotation, av);
 				av.visitEnd();
@@ -136,56 +141,163 @@ public class ExtensionProxyFactory {
 		}
 	}
 
+	private static List<ParameterizedType> getInterfacesAndGenericSuperclasses(Class<?> toCheck, List<ParameterizedType> fromChildren, Set<Class<?>> remainingContracts) {
+		if(toCheck == Object.class) {
+			return fromChildren;
+		}
+		
+		for (java.lang.reflect.Type type : toCheck.getGenericInterfaces()) {
+			if(type instanceof ParameterizedType) {
+				ParameterizedType pt = (ParameterizedType) type;
+				if(remainingContracts.remove(pt.getRawType())) {
+					fromChildren.add(pt);
+				}
+			}
+		}
+		
+		java.lang.reflect.Type genericSuperclass = toCheck.getGenericSuperclass();
+		if(genericSuperclass instanceof ParameterizedType) {
+			fromChildren.add((ParameterizedType) genericSuperclass);
+		}
+		
+		return getInterfacesAndGenericSuperclasses(toCheck.getSuperclass(), fromChildren, remainingContracts);
+	}
+	
+	private static Predicate<TypeVariable<?>> isUsedInTypeInfo(Map<String, ParameterizedType> typeInfo) {
+		return tv -> typeInfo.values().stream()
+				.flatMap(ExtensionProxyFactory::toTypeVariables)
+				.anyMatch(t -> tv.getName().equals(t.getName()));
+	}
+	
 	/**
 	 * @param typeInfo
 	 * @return
 	 */
-	private static String generateGenericClassSignature(Map<String, ParameterizedType> typeInfo, List<Class<?>> contracts) {
+	private static String generateGenericClassSignature(Class<?> delegateClazz, Map<String, ParameterizedType> typeInfo, List<Class<?>> contracts) {
 		if(typeInfo.isEmpty()) {
 			return null;
 		}
 		
-		SignatureWriter sw = new SignatureWriter();
-		// Handle parent
-		SignatureVisitor sv = sw.visitSuperclass();
+		SignatureWriter writer = new SignatureWriter();
+		
+		// Handle class
+		Arrays.stream(delegateClazz.getTypeParameters())
+			.filter(isUsedInTypeInfo(typeInfo))
+			.forEach(tv -> {
+				writer.visitFormalTypeParameter(tv.getName());
+				SignatureVisitor cb = writer.visitClassBound();
+				Arrays.stream(tv.getBounds()).forEach(b -> visitTypeParameter(b, cb, delegateClazz, typeInfo));
+				cb.visitEnd();
+			});
+		
+		// Generated class extends Object
+		SignatureVisitor sv = writer.visitSuperclass();
 		sv.visitClassType(OBJECT_INTERNAL_NAME);
 		sv.visitEnd();
 		
 		// Handle interfaces
 		for(Class<?> contract : contracts) {
 			if(typeInfo.containsKey(contract.getName())) {
-				SignatureVisitor iv = sw.visitInterface();
+				Class<?> directDeclarer = delegateClazz;
+				check: for(;;) {
+					for(Class<?> iface : directDeclarer.getInterfaces()) {
+						if(iface == contract) {
+							break check;
+						}
+					}
+					directDeclarer = directDeclarer.getSuperclass();
+					if(directDeclarer == Object.class) {
+						throw new IllegalArgumentException("The contract " + contract + " is not implemented in the hierarchy");
+					}
+				}
+				SignatureVisitor iv = writer.visitInterface();
 				iv.visitClassType(Type.getInternalName(contract));
 				for(java.lang.reflect.Type t : typeInfo.get(contract.getName()).getActualTypeArguments()) {
-					visitTypeParameter(t, iv);
+					if(TypeVariable.class.isInstance(t)) {
+						visitTypeParameter(t, iv, directDeclarer, typeInfo);
+					} else {
+						SignatureVisitor tav = iv.visitTypeArgument(SignatureVisitor.INSTANCEOF);
+						visitTypeParameter(t, tav, directDeclarer, typeInfo);
+						tav.visitEnd();
+					}
 				}
 				iv.visitEnd();
 			}
 		}
 		 
-		sw.visitEnd();
-		return sw.toString();
+		writer.visitEnd();
+		return writer.toString();
+	}
+	
+	private static java.lang.reflect.Type getPossibleReifiedTypeFor(TypeVariable<?> tv, Class<?> directDeclarer, Map<String, ParameterizedType> typeInfo) {
+		ParameterizedType pt = typeInfo.get(directDeclarer.getName());
+		if(pt != null) {
+			TypeVariable<?>[] decParams = directDeclarer.getTypeParameters();
+			for (int i = 0; i < decParams.length; i++) {
+				TypeVariable<?> decTv = decParams[i];
+				if(decTv.getName().equals(tv.getName())) {
+					return pt.getActualTypeArguments()[i];
+				}
+			}
+		} 
+		return tv;
+	}
+	
+	private static Stream<TypeVariable<?>> toTypeVariables(java.lang.reflect.Type t) {
+		if(t instanceof Class<?>) {
+			return Stream.empty();
+		} else if (t instanceof TypeVariable<?>) {
+			return Stream.of((TypeVariable<?>)t);
+		} else if (t instanceof ParameterizedType) {
+			return Arrays.stream(((ParameterizedType) t).getActualTypeArguments())
+					.flatMap(ExtensionProxyFactory::toTypeVariables);
+		} else {
+			throw new IllegalArgumentException("Unkown type " + t.getClass());
+		}
 	}
 
-	private static void visitTypeParameter(java.lang.reflect.Type t, SignatureVisitor sv) {
-		SignatureVisitor pv = sv.visitTypeArgument('=');
+	private static void visitTypeParameter(java.lang.reflect.Type t, SignatureVisitor sv, Class<?> directDeclarer, Map<String, ParameterizedType> typeInfo) {
 		if(t instanceof Class<?>) {
 			Class<?> clazz = (Class<?>) t;
 			if(clazz.isPrimitive()) {
-				pv.visitBaseType(Type.getDescriptor(clazz).charAt(0));
+				sv.visitBaseType(Type.getDescriptor(clazz).charAt(0));
 			} else if (clazz.isArray()) {
-				SignatureVisitor av = pv.visitArrayType();
-				visitTypeParameter(clazz.getComponentType(), av);
+				SignatureVisitor av = sv.visitArrayType();
+				visitTypeParameter(clazz.getComponentType(), av, directDeclarer, typeInfo);
 				av.visitEnd();
 			} else {
-				pv.visitClassType(Type.getInternalName(clazz));
+				sv.visitClassType(Type.getInternalName(clazz));
 			}
 		} else if (t instanceof ParameterizedType){
 			ParameterizedType pt = (ParameterizedType) t;
-			pv.visitClassType(Type.getInternalName((Class<?>)pt.getRawType()));
-			Arrays.stream(pt.getActualTypeArguments()).forEach(ta -> visitTypeParameter(ta, pv));
-		}
-		pv.visitEnd();
+			sv.visitClassType(Type.getInternalName((Class<?>)pt.getRawType()));
+			Arrays.stream(pt.getActualTypeArguments()).forEach(ta -> visitTypeParameter(ta, sv, directDeclarer, typeInfo));
+		} else if (t instanceof TypeVariable<?>) {
+			TypeVariable<?> tv = (TypeVariable<?>) t;
+			t = getPossibleReifiedTypeFor((TypeVariable<?>)t, directDeclarer, typeInfo);
+			if(t == tv) {
+				sv.visitTypeArgument(SignatureVisitor.INSTANCEOF).visitTypeVariable(tv.getName());
+			} else {
+				SignatureVisitor tav = sv.visitTypeArgument(SignatureVisitor.INSTANCEOF);
+				visitTypeParameter(t, tav, directDeclarer, typeInfo);
+				tav.visitEnd();
+			}
+		} else if (t instanceof WildcardType) {
+			WildcardType wt = (WildcardType) t;
+			SignatureVisitor tav;
+			java.lang.reflect.Type[] types;
+			if(wt.getLowerBounds().length > 0) {
+				tav = sv.visitTypeArgument(SignatureVisitor.SUPER);
+				types = wt.getLowerBounds();
+			} else {
+				tav = sv.visitTypeArgument(SignatureVisitor.EXTENDS);
+				types = wt.getUpperBounds();
+			}
+			Arrays.stream(types).forEach(ty -> visitTypeParameter(ty, tav, directDeclarer, typeInfo));
+			tav.visitEnd();
+		} else {
+			throw new IllegalArgumentException("Unhandled generic type " + t.getClass() + " " + t.toString());
+ 		}
 	}
 	
 	private static void visitAnnotationMembers(Annotation a, AnnotationVisitor av) throws IllegalAccessException, IllegalArgumentException, InvocationTargetException {

--- a/org.eclipse.osgitech.rest/src/test/java/org/eclipse/osgitech/rest/proxy/ExtensionProxyTest.java
+++ b/org.eclipse.osgitech.rest/src/test/java/org/eclipse/osgitech/rest/proxy/ExtensionProxyTest.java
@@ -29,21 +29,25 @@ import java.io.OutputStream;
 import java.lang.annotation.Annotation;
 import java.lang.reflect.ParameterizedType;
 import java.lang.reflect.Type;
+import java.lang.reflect.TypeVariable;
+import java.util.ArrayList;
 import java.util.Arrays;
+import java.util.Collections;
 import java.util.List;
 import java.util.function.Supplier;
 
+import org.junit.jupiter.api.Test;
+
+import jakarta.ws.rs.NotFoundException;
 import jakarta.ws.rs.Path;
 import jakarta.ws.rs.WebApplicationException;
 import jakarta.ws.rs.core.MediaType;
 import jakarta.ws.rs.core.MultivaluedMap;
 import jakarta.ws.rs.core.Response;
+import jakarta.ws.rs.ext.ContextResolver;
 import jakarta.ws.rs.ext.ExceptionMapper;
 import jakarta.ws.rs.ext.MessageBodyReader;
 import jakarta.ws.rs.ext.MessageBodyWriter;
-
-import org.eclipse.osgitech.rest.proxy.ExtensionProxyFactory;
-import org.junit.jupiter.api.Test;
 
 /**
  * 
@@ -119,6 +123,204 @@ public class ExtensionProxyTest {
 		
 		assertEquals(814, ((ExceptionMapper)instance)
 				.toResponse(new OutOfMemoryError()).getStatus());
+		
+	}
+	
+	@SuppressWarnings("unchecked")
+	@Test
+	public void testParameterExceptionMapper() throws Exception {
+		
+		ExceptionMapper<?> em = new ParameterExceptionMapper<>();
+		
+		Class<?> proxyClazz = pcl.define("test.ExceptionMapper", em, 
+				singletonList(ExceptionMapper.class));
+		
+		assertTrue(ExceptionMapper.class.isAssignableFrom(proxyClazz));
+		
+		TypeVariable<?>[] typeParameters = proxyClazz.getTypeParameters();
+		assertEquals(1, typeParameters.length);
+		assertTrue(TypeVariable.class.isInstance(typeParameters[0]));
+		assertEquals("T", typeParameters[0].toString());
+		Type[] bounds = ((TypeVariable<?>)typeParameters[0]).getBounds();
+		assertEquals(1, bounds.length);
+		assertEquals(Throwable.class, bounds[0]);
+		
+		Type[] genericInterfaces = proxyClazz.getGenericInterfaces();
+		
+		assertEquals(1, genericInterfaces.length);
+		assertTrue(genericInterfaces[0] instanceof ParameterizedType);
+		ParameterizedType pt = (ParameterizedType) genericInterfaces[0];
+		assertEquals(ExceptionMapper.class, pt.getRawType());
+		assertTrue(TypeVariable.class.isInstance(pt.getActualTypeArguments()[0]));
+		assertEquals("T", pt.getActualTypeArguments()[0].toString());
+	
+		
+		Object instance = proxyClazz.getConstructor(Supplier.class)
+			.newInstance((Supplier<?>) () -> em);
+		
+		assertEquals(451, ((ExceptionMapper<Throwable>)instance)
+				.toResponse(new RuntimeException()).getStatus());
+		
+	}
+
+	@SuppressWarnings("unchecked")
+	@Test
+	public void testChildReifiedExceptionMapper() throws Exception {
+		
+		ExceptionMapper<WebApplicationException> em = new ChildReifiedExceptionMapper();
+		
+		Class<?> proxyClazz = pcl.define("test.ExceptionMapper", em, 
+				singletonList(ExceptionMapper.class));
+		
+		assertTrue(ExceptionMapper.class.isAssignableFrom(proxyClazz));
+		
+		TypeVariable<?>[] typeParameters = proxyClazz.getTypeParameters();
+		assertEquals(0, typeParameters.length);
+		
+		Type[] genericInterfaces = proxyClazz.getGenericInterfaces();
+		
+		assertEquals(1, genericInterfaces.length);
+		assertTrue(genericInterfaces[0] instanceof ParameterizedType);
+		ParameterizedType pt = (ParameterizedType) genericInterfaces[0];
+		assertEquals(ExceptionMapper.class, pt.getRawType());
+		assertEquals(WebApplicationException.class, pt.getActualTypeArguments()[0]);
+		
+		
+		Object instance = proxyClazz.getConstructor(Supplier.class)
+				.newInstance((Supplier<?>) () -> em);
+		
+		assertEquals(404, ((ExceptionMapper<WebApplicationException>)instance)
+				.toResponse(new NotFoundException()).getStatus());
+		
+	}
+	
+	@SuppressWarnings("unchecked")
+	@Test
+	public void testParameterContext() throws Exception {
+		
+		ContextResolver<? extends CharSequence> cr = new ParameterContext<>();
+		
+		Class<?> proxyClazz = pcl.define("test.ContextResolver", cr, 
+				singletonList(ContextResolver.class));
+		
+		TypeVariable<?>[] typeParameters = proxyClazz.getTypeParameters();
+		assertEquals(1, typeParameters.length);
+		assertTrue(TypeVariable.class.isInstance(typeParameters[0]));
+		assertEquals("T", typeParameters[0].toString());
+		Type[] bounds = ((TypeVariable<?>)typeParameters[0]).getBounds();
+		assertEquals(1, bounds.length);
+		assertEquals(Object.class, bounds[0]);
+		
+		assertTrue(ContextResolver.class.isAssignableFrom(proxyClazz));
+		Type[] genericInterfaces = proxyClazz.getGenericInterfaces();
+		
+		assertEquals(1, genericInterfaces.length);
+		assertTrue(genericInterfaces[0] instanceof ParameterizedType);
+		ParameterizedType pt = (ParameterizedType) genericInterfaces[0];
+		assertEquals(ContextResolver.class, pt.getRawType());
+		assertTrue(TypeVariable.class.isInstance(pt.getActualTypeArguments()[0]));
+		assertEquals("T", pt.getActualTypeArguments()[0].toString());
+		
+		
+		Object instance = proxyClazz.getConstructor(Supplier.class)
+				.newInstance((Supplier<?>) () -> cr);
+		
+		assertEquals("", ((ContextResolver<? extends CharSequence>)instance)
+				.getContext(String.class));
+		
+	}
+
+	@SuppressWarnings("unchecked")
+	@Test
+	public void testExtendsParameterContext() throws Exception {
+		
+		ContextResolver<List<Integer>> cr = new ExtendsParameterContext<>();
+		
+		Class<?> proxyClazz = pcl.define("test.ContextResolver", cr, 
+				singletonList(ContextResolver.class));
+		
+		TypeVariable<?>[] typeParameters = proxyClazz.getTypeParameters();
+		assertEquals(1, typeParameters.length);
+		assertTrue(TypeVariable.class.isInstance(typeParameters[0]));
+		assertEquals("R", typeParameters[0].toString());
+		Type[] bounds = ((TypeVariable<?>)typeParameters[0]).getBounds();
+		assertEquals(1, bounds.length);
+		assertEquals(Number.class, bounds[0]);
+		
+		assertTrue(ContextResolver.class.isAssignableFrom(proxyClazz));
+		Type[] genericInterfaces = proxyClazz.getGenericInterfaces();
+		
+		assertEquals(1, genericInterfaces.length);
+		assertTrue(genericInterfaces[0] instanceof ParameterizedType);
+		ParameterizedType pt = (ParameterizedType) genericInterfaces[0];
+		assertEquals(ContextResolver.class, pt.getRawType());
+		assertTrue(ParameterizedType.class.isInstance(pt.getActualTypeArguments()[0]));
+		assertEquals("java.util.List<R>", pt.getActualTypeArguments()[0].toString());
+		
+		
+		Object instance = proxyClazz.getConstructor(Supplier.class)
+				.newInstance((Supplier<?>) () -> cr);
+		
+		assertEquals(Collections.emptyList(), ((ContextResolver<List<Integer>>)instance)
+				.getContext(String.class));
+		
+	}
+
+	@SuppressWarnings("unchecked")
+	@Test
+	public void testChildExtendsParameterContext() throws Exception {
+		
+		ContextResolver<List<Double>> cr = new ChildExtendsParameterContext();
+		
+		Class<?> proxyClazz = pcl.define("test.ContextResolver", cr, 
+				singletonList(ContextResolver.class));
+		
+		TypeVariable<?>[] typeParameters = proxyClazz.getTypeParameters();
+		assertEquals(0, typeParameters.length);
+		
+		assertTrue(ContextResolver.class.isAssignableFrom(proxyClazz));
+		Type[] genericInterfaces = proxyClazz.getGenericInterfaces();
+		
+		assertEquals(1, genericInterfaces.length);
+		assertTrue(genericInterfaces[0] instanceof ParameterizedType);
+		ParameterizedType pt = (ParameterizedType) genericInterfaces[0];
+		assertEquals(ContextResolver.class, pt.getRawType());
+		assertTrue(ParameterizedType.class.isInstance(pt.getActualTypeArguments()[0]));
+		assertEquals("java.util.List<java.lang.Double>", pt.getActualTypeArguments()[0].toString());
+		
+		
+		Object instance = proxyClazz.getConstructor(Supplier.class)
+				.newInstance((Supplier<?>) () -> cr);
+		
+		assertEquals(Collections.singletonList(42.0d), ((ContextResolver<List<Double>>)instance)
+				.getContext(String.class));
+		
+	}
+
+	@SuppressWarnings("unchecked")
+	@Test
+	public void testWildcardParameterContext() throws Exception {
+		
+		ContextResolver<List<? super Integer>> cr = new WildcardParameterContext();
+		
+		Class<?> proxyClazz = pcl.define("test.ContextResolver", cr, 
+				singletonList(ContextResolver.class));
+		assertTrue(ContextResolver.class.isAssignableFrom(proxyClazz));
+		Type[] genericInterfaces = proxyClazz.getGenericInterfaces();
+		
+		assertEquals(1, genericInterfaces.length);
+		assertTrue(genericInterfaces[0] instanceof ParameterizedType);
+		ParameterizedType pt = (ParameterizedType) genericInterfaces[0];
+		assertEquals(ContextResolver.class, pt.getRawType());
+		assertTrue(ParameterizedType.class.isInstance(pt.getActualTypeArguments()[0]));
+		assertEquals("java.util.List<? super java.lang.Integer>", pt.getActualTypeArguments()[0].toString());
+		
+		
+		Object instance = proxyClazz.getConstructor(Supplier.class)
+				.newInstance((Supplier<?>) () -> cr);
+		
+		assertEquals(Collections.emptyList(), ((ContextResolver<List<Integer>>)instance)
+				.getContext(String.class));
 		
 	}
 	
@@ -225,6 +427,70 @@ public class ExtensionProxyTest {
 		public Response toResponse(Throwable exception) {
 			// What's an 8xx response?!?
 			return Response.status(814).build();
+		}
+		
+	}
+
+	public static class ParameterExceptionMapper<T extends Throwable> implements ExceptionMapper<T> {
+		
+		/* 
+		 * (non-Javadoc)
+		 * @see jakarta.ws.rs.ext.ExceptionMapper#toResponse(java.lang.Throwable)
+		 */
+		@Override
+		public Response toResponse(T exception) {
+			// Unavailable for legal reasons
+			return Response.status(451).build();
+		}
+		
+	}
+	
+	public static class ChildReifiedExceptionMapper extends ParameterExceptionMapper<WebApplicationException> {
+
+		@Override
+		public Response toResponse(WebApplicationException exception) {
+			return Response.status(exception.getResponse().getStatus()).build();
+		}
+		
+	}
+
+	public static class ParameterContext<T> implements ContextResolver<T> {
+
+		@SuppressWarnings("unchecked")
+		@Override
+		public T getContext(Class<?> type) {
+			try {
+				return (T) type.getConstructor().newInstance();
+			} catch (Exception e) {
+				throw new RuntimeException(e);
+			}
+		}
+		
+	}
+
+	public static class ExtendsParameterContext<R extends Number> implements ContextResolver<List<R>> {
+		
+		@Override
+		public List<R> getContext(Class<?> type) {
+			return new ArrayList<>();
+		}
+		
+	}
+
+	public static class ChildExtendsParameterContext extends ExtendsParameterContext<Double> {
+		
+		@Override
+		public List<Double> getContext(Class<?> type) {
+			return Collections.singletonList(42.0d);
+		}
+		
+	}
+
+	public static class WildcardParameterContext implements ContextResolver<List<? super Integer>> {
+		
+		@Override
+		public List<? super Integer> getContext(Class<?> type) {
+			return new ArrayList<>();
 		}
 		
 	}

